src/java/com/palantir/cassandra/utils/FileParser.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.cassandra.utils;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardCopyOption;
+import java.nio.file.StandardOpenOption;
+import java.util.Optional;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+public class FileParser<T>
+{
+    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+
+    private final Path path;
+
+    private final TypeReference<T> cls;
+
+    public FileParser(Path path, TypeReference<T> cls)
+    {
+        this.path = path;
+        this.cls = cls;
+    }
+
+    public Optional<T> read() throws IOException
+    {
+        File file = path.toFile();
+        if (Files.isRegularFile(path) && file.length() > 0)
+        {
+            return Optional.of(OBJECT_MAPPER.readValue(file, cls));
+        }
+        else
+        {
+            return Optional.empty();
+        }
+    }
+
+    public void write(T value) throws IOException
```

Comment:
I think we should be good not making this synchronized since `CleanupStateTracker` functions are are synchronized and `VolumesIntegrityCheck` does writes sequentially but might be worth double checking

Reply:
From reading the code, I believe it should be fine. The only edge case I see is if multiple threads try to call `StorageService.init` which will hit `prepareToJoin` codepath multiple time. Is there any reason we should account for this?